### PR TITLE
Sprint 2: Login, Registration, and Main Pages Implementation

### PR DESCRIPTION
**Pull request by FedAliaks**
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint%232.md
2. Screenshot:
   image
3. Deploy: https://hurricane-bookstore.netlify.app/
4. Done 20.05.2024 / deadline 20.05.2024
5. Cross-chet: 315 / 315, Mentor evaluation: 60/60

#### Input Validation (40 points)
- [x]	(20 points) Implement client-side validation for the login form, including email and password fields. RSS-ECOMM-2_01
- [x]	(20 points) Display clear error messages indicating any validation issues, such as an improperly formatted email. RSS-ECOMM-2_02
#### ntegration with Authentication Service (45 points)
- [x]	(25 points) Integrate the login form with a chosen authentication service (CommerceTools) to handle user authentication. RSS-ECOMM-2_03
- [x]	(20 points) Implement error handling for failed authentication attempts, such as incorrect email or password, and display user-friendly error messages. RSS-ECOMM-2_04
####  Redirection (30 points)
- [x]	(15 points) Redirect users to the application's main page upon successful login. RSS-ECOMM-2_05
- [x]	(15 points) Redirect users who are already logged in to the main page if they try to access the login page. RSS-ECOMM-2_06
####  Handle Authentication Token (10 points)
- [x]	(10 points) Obtain the authentication token securely after a successful login attempt by sending a request to the token endpoint (e.g., https://auth.europe-west1.gcp.commercetools.com/oauth/project_key/customers/token), allowing for seamless user authentication across the application. RSS-ECOMM-2_07
####  Navigation to Registration Page (5 points)
- [x]	(5 points) Add a button or link on the login page that allows users to navigate to the registration page. RSS-ECOMM-2_08

####  Input Validation (45 points)
- [x]	(25 points) Implement client-side validation for all required fields in the registration form, such as email, password, first nae, last name, date of birth, and address fields (e.g., street, city, postal code, and country) for proper use with CommerceTools. RSS-ECOMM-2_09
- [x]	(20 points) Display clear error messages indicating any validation issues, such as an improperly formatted email or a weak password. RSS-ECOMM-2_10
####  Integration with Authentication Service (25 points)
- [x]	(10 points) Integrate the registration form with a chosen authentication service, such as commercetools, to handle user registration. RSS-ECOMM-2_11
- [x]	(15 points) Implement error handling for failed registration attempts, and display user-friendly error messages. RSS-ECOMM-2_12
####  State Management, Automatic Login, and Redirection (15 points)
- [x]	15 points) Redirect users to the application's main page upon successful account creation and automatic login. RSS-ECOMM-2_13
####  Integration with commercetools for User Profiles and Addresses (30 points)
- [x]	(15 points) Allow users to set a default address during registration. RSS-ECOMM-2_14
- [x]	15 points) Enable users to select different billing and shipping addresses or choose a single address for both billing and shipping during the registration process. RSS-ECOMM-2_15
####  Navigation to Login Page (5 points)
- [x]	5 points) Add a button or link on the registration page that allows users to navigate to the login page. RSS-ECOMM-2_16

####  Centralized Navigation (10 points)
- [x]	(5 points) Add links to all the functional pages of the application on the main page. These should include, but are not limited to, the login and registration pages. RSS-ECOMM-2_17
- [x]	(5 points) Each link should redirect the user correctly to the corresponding page without any errors. RSS-ECOMM-2_18

###  Routing Implementation (30 points - Total)
- [x]	(15 points) Implement routing for navigation between login, registration, and main pages. RSS-ECOMM-2_19
- [x]	 (15 points) Implement a 404 (Not Found) page for invalid route requests. RSS-ECOMM-2_20

###   Evaluation Criteria for Header (25 points - Total)
- [x]	(15 points) Navigation to login and registration pages for unauthorized users RSS-ECOMM-2_21
- [x]	(5 points) Ability to access the main page for all users RSS-ECOMM-2_22
- [x]	(5 points) Logout functionality for authorized users RSS-ECOMM-2_23

###  Mentor Evaluation Criteria (45 /60):
- [x]	Code Review Participation (10 points): Students actively participate in code reviews, providing constructive feedback and suggestions to their peers.
- [x]	No Merges Without Approvals (10 points): Pull requests are not merged without approvals from other team members, ensuring the quality of code contributions.
- [x]	Pull Request Descriptions (10 points): Pull requests include detailed descriptions, outlining the changes made and the purpose of the changes, facilitating better understanding and review.
- [x]	Task Board Management (10 points): Students maintain and regularly update a task board (e.g., Trello, Jira), ensuring that tasks are moved according to their progress and completion.
- [x]	Active Participation in Development Discussions (5 points): Students take an active role in discussing the development process, sharing ideas, and collaborating on solutions and improvements.
- [x] Unit Test Coverage (15 points): The codebase has a minimum of 25% unit test coverage, ensuring the reliability and robustness of the implemented features.

### CODE REVIEW

**Review by DaryaAniskevich — APPROVED** [review]

File: rs-commerce/src/types/registrationTypes.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  | 'setShippingDefaultAddress'
+  | 'setBillingDefaultAddress';
+
+export type CheckboxRegistrationType = {
+  content: string;
+  htmlFor: HtmlForCheckboxType;
+};
+
+export enum TypeFields {
+  /*     dateOfBirth = 'dateOfBirth', */
```
Comment (sergioivanov008) [inline rs-commerce/src/types/registrationTypes.ts]: It's better to remove it

File: rs-commerce/src/views/registration/components/RegistrationForm/components/AddressRegistration/components/AddressComponents/AddressComponents.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { AddressType, HtmlForType, InputType } from 'types/registrationTypes';
+import CountryInput from '../../../InputRegistration/CountryInput';
+import InputRegistration from '../../../InputRegistration/InputRegistration';
+import AddressTitleComponent from './components/AddresTitleComponent/AddresTitleComponent';
+import classes from './style.module.css';
+
+/* let typeComponent: 'billing' | 'shipping' = 'billing'; */
```
Comment (sergioivanov008) [inline rs-commerce/src/views/registration/components/RegistrationForm/components/AddressRegistration/components/AddressComponents/AddressComponents.tsx]: It's better to remove it

File: rs-commerce/src/views/registration/components/RegistrationForm/components/InputRegistration/InputRegistration.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    inputSize = `${classes['registration__input']} ${classes['registration__input_small']}`;
+  }
+
+  const [errorContent, setErrorContent] = useState('');
+
+  /*   function clearField(e: React.FocusEvent<HTMLInputElement, Element>) {
+    e.target.value = '';
+    setErrorContent('');
+    dispatch(registrationFormActions[DispatchObj[htmlFor]](''));
+  } */
```
Comment (sergioivanov008) [inline rs-commerce/src/views/registration/components/RegistrationForm/components/InputRegistration/InputRegistration.tsx]: It's better to remove it

File: rs-commerce/src/views/registration/components/RegistrationForm/components/InputRegistration/InputRegistration.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      <label htmlFor={htmlFor}>
+        <div className={classes['input__block']}>
+          <input
+            className={inputSize}
+            id={htmlFor}
+            type={type}
+            disabled={addDisabled()}
+            placeholder={placeholder}
+            onChange={(e) => checkValue(e)}
+            /*             onFocus={(e) => clearField(e)} */
```
Comment (sergioivanov008) [inline rs-commerce/src/views/registration/components/RegistrationForm/components/InputRegistration/InputRegistration.tsx]: It's better to remove it

File: rs-commerce/src/views/registration/components/RegistrationForm/RegistrationForm.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+
+  const ButtonRegistrationClick = async (): Promise<void> => {
+    const bodyRegistration: RegistrationCustomerType = {
+      email,
+      password,
+    };
+
+    // Create the customer and output the Customer ID
```
Comment (sergioivanov008) [inline rs-commerce/src/views/registration/components/RegistrationForm/RegistrationForm.tsx:50]: It's better to remove it

**Review by sergioivanov008 — COMMENTED** [review]
Some lines better to remove

**Review by sergioivanov008 — APPROVED** [review]

**Pull request merged**